### PR TITLE
More monitor

### PR DESCRIPTION
**Pull request by jonasfj**
I'll enable the lint test and fix issues with it later...

### CODE REVIEW

File: src/deadlineresolver.js [hunk 1]
```diff
@@ -80,8 +81,13 @@ class DeadlineResolver extends events.EventEmitter {
...
     // Create promise that we're done looping
-    this.done = Promise.all(loops).catch((err) => {
-      this.emit('error', err); // This should crash the process
+    this.done = Promise.all(loops).catch(async (err) => {
+      console.log("Crashing the process: %s, as json: %j", err, err);
+      // TODO: use this.monitor.reportError(err); when PR lands:
+      // https://github.com/taskcluster/taskcluster-lib-monitor/pull/27
+      await this.monitor.reportError(err, 'error', {}, true);
+      // Crash the process
+      process.exit(1);
```
Comment (djmitche) [inline src/deadlineresolver.js]: It seems like this behavior should be factored out to somewhere common, perhaps with tests?

File: src/artifacts.js [hunk 2]
```diff
@@ -436,8 +438,9 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   }
 
   // We should never arrive here
-  debug("[alert-operator] Unknown artifact storageType from table storage: %s",
-        artifact.storageType);
+  let err = new Error('Unknown artifact storageType: ' + artifacts.storageType);
+  err.artifact = artifact.json();
+  this.monitor.reportError(err);
```
Comment (gregarndt) [inline src/artifacts.js]: should wait for this promise to resolve?

File: src/bucket.js [hunk 3]
```diff
@@ -23,11 +24,14 @@ var Bucket = function(options) {
   assert(options.credentials, "credentials must be specified");
   assert(!options.bucketCDN || typeof(options.bucketCDN) === 'string',
          "Expected bucketCDN to be a hostname or empty string for none");
+  assert(options.monitor,     "options.monitor is required");
```
Comment (gregarndt) [inline src/bucket.js]: nit, this error doesn't is a little different than the others... perhaps just "monitor must be given" ?  I'm using the bucket and credentials assertions above as an example.

File: src/bucket.js [hunk 4]
```diff
@@ -149,11 +153,10 @@ Bucket.prototype.setupCORS = async function() {
...
       return;
     }
-  }
-  catch (err) {
+  } catch (err) {
     // Failed to fetch CORS, ignoring issue for now
-    debug("[alert-operator] Failed to fetch CORS, err: %s, JSON: %j",
-          err, err, err.stack);
+    err.note = 'Failed to fetch CORS in bucket.js';
+    this.monitor.reportError(err, 'warning');
```
Comment (gregarndt) [inline src/bucket.js]: need to wait for this to complete?

File: src/claimresolver.js [hunk 5]
```diff
@@ -28,7 +28,7 @@ const RESOLVED_STATES = [
  * ignored as it implies a failed request (which wasn't retried) or a run
  * that was reclaimed.
  */
-class ClaimResolver extends events.EventEmitter {
+class ClaimResolver {
```
Comment (gregarndt) [inline src/claimresolver.js]: if this no longer extends EventEmitter, do you need to import events anymore?

File: src/claimresolver.js [hunk 6]
```diff
@@ -57,12 +56,14 @@ class ClaimResolver extends events.EventEmitter {
            "Expected pollingDelay to be a number");
     assert(typeof(options.parallelism) === 'number',
            "Expected parallelism to be a number");
+    assert(options.monitor !== null, 'options.monitor required!');
```
Comment (gregarndt) [inline src/claimresolver.js]: same comment I gave with the other assertion

File: src/claimresolver.js [hunk 7]
```diff
@@ -102,13 +108,14 @@ class ClaimResolver extends events.EventEmitter {
...
         // Don't let a single task error break the loop, it'll be retried later
         // as we don't remove message unless they are handled
-        return this.handleMessage(message).catch((err) => {
-          debug("[alert-operator] Failed to handle message: %j" +
-                ", with err: %s, as JSON: %j", message, err, err, err.stack);
-        });
+        try {
+          await this.handleMessage(message);
+        } catch (err) {
+          this.monitor.reportError(err, 'warning');
```
Comment (gregarndt) [inline src/claimresolver.js]: Before you were including the message in the alert-operator statement, do you not want to include that in the error you're reporting?

File: src/claimresolver.js [hunk 8]
```diff
@@ -191,8 +200,11 @@ class ClaimResolver extends events.EventEmitter {
 
       // If the run isn't the last run, then something is very wrong
       if (task.runs.length - 1 !== runId) {
-        debug("[alert-operator] running runId: %s, resolved exception, " +
-              "but it wasn't the last run! taskId: ", runId, taskId);
+        let err = new Error('Running runId: ' + runId + ' resolved exception,' +
+                            'but it was not the last run! taskId: ' + taskId);
+        err.taskId = taskId;
+        err.runId = runId;
+        this.monitor.reportError(err);
```
Comment (gregarndt) [inline src/claimresolver.js]: await?

File: src/data.js [hunk 9]
```diff
@@ -316,9 +317,12 @@ Artifact.prototype.remove = function(ignoreError) {
...
     } else {
-      debug("[alert-operator] Expiring artifact with bucket: %s, which isn't " +
-            "configured for use. Please investigate taskId: %s, runId: %s",
-            this.details.bucket, this.taskId, this.runId);
+      let err = new Error("Expiring artifact with bucket which isn't " +
+                          "configured for use. Please investigate!");
+      err.bucket  = this.details.bucket;
+      err.taskId  = this.taskId;
+      err.runId   = this.runId;
+      this.monitor.reportError(err);
```
Comment (gregarndt) [inline src/data.js]: await?

File: src/data.js [hunk 10]
```diff
@@ -316,9 +317,12 @@ Artifact.prototype.remove = function(ignoreError) {
     } else if (this.details.bucket === this.privateBucket.bucket) {
       deleted = this.privateBucket.deleteObject(this.details.prefix);
     } else {
-      debug("[alert-operator] Expiring artifact with bucket: %s, which isn't " +
-            "configured for use. Please investigate taskId: %s, runId: %s",
-            this.details.bucket, this.taskId, this.runId);
+      let err = new Error("Expiring artifact with bucket which isn't " +
```
Comment (gregarndt) [inline src/data.js]: nit: you intermingle single and double quotes when you deal with strings, I suppose the linter stuff will pick up on this if there is the right rule for it.

File: src/data.js [hunk 11]
```diff
@@ -329,9 +333,12 @@ Artifact.prototype.remove = function(ignoreError) {
...
     if (this.details.container !== this.blobStore.container) {
-      debug("[alert-operator] Expiring artifact with container: %s, which " +
-            "configured for use. Please investigate taskId: %s, runId: %s",
-            this.details.container, this.taskId, this.runId);
+      let err = new Error("Expiring artifact with container which isn't " +
+                          "configured for use. Please investigate!");
+      err.container = this.details.container;
+      err.taskId    = this.taskId;
+      err.runId     = this.runId;
+      this.monitor.reportError(err);
```
Comment (gregarndt) [inline src/data.js]: ok, so I think that this should be used with "await" but I'll refrain from commenting on it anymore just to reduce the noise.

File: src/deadlineresolver.js (hunk 1 above)
Comment (gregarndt) [inline src/deadlineresolver.js]: +1, especially since this piece I think has bitten us twice now.

**Comment by gregarndt** [comment]
Took a look and I think the larged chunk of comments is regarding awaiting for the resolution of the reportError() promise.

File: src/claimresolver.js (hunk 5 above)
Comment (jonasfj) [inline src/claimresolver.js]: good point

File: src/data.js (hunk 10 above)
Comment (jonasfj) [inline src/data.js]: linter PR will fix this...

**Comment by jonasfj** [comment]
I have mixed feelings with regards to waiting for the `monitor.reportError` to resolve...

I see error reporting as a best effort service... So I'm not sure there is any need to wait for it in most cases... I do it if the error is really critical...

But it's mostly for cases where I want to shutdown after reporting the error...

**Comment by jonasfj** [comment]
Does that make sense?

**Comment by gregarndt** [comment]
makes sense, thanks.

File: src/deadlineresolver.js (hunk 1 above)
Comment (jonasfj) [inline src/deadlineresolver.js]: yeah, that would be nice... It won't be in this PR though...
Ideally, some generic way of constructing a process that polls from azure queue and operates on the messages...

**Pull request merged**